### PR TITLE
Support for pre() and post() hook for FinancialTrxn

### PR DESCRIPTION
**Pull request by pradpnayak**
Overview
----------------------------------------
Not sure if its a right thing to allow to alter the financial trxn params using pre() and post() hook but I feel it should be allowed so that one can change the params like to/from_financial_account or create additional transactions in post() hook etc. Thoughts?

(Note: This PR is work in progress)

If approved will update the PR with
1. Move CRM/Core/BAO/FinancialTrxn.php to CRM/Financial/BAO/FinancialTrxn.
2. Replace class in all code from CRM_Core_BAO_FinancialTrxn to CRM_Financial_BAO_FinancialTrxn
3. Create empty file CRM/Core/BAO/FinancialTrxn.php which extends CRM_Financial_BAO_FinancialTrxn (not sure how to write deprecated warning!!!!)

### CODE REVIEW

**Comment by mattwire** [comment]
@pradpnayak You should already be able to listen for `civi.dao.pre[Update|Insert]`, `civi.dao.post[Update|Insert]` symfony events?

**Comment by pradpnayak** [comment]
> @pradpnayak You should already be able to listen for `civi.dao.pre[Update|Insert]`, `civi.dao.post[Update|Insert]` symfony events?

Yes you can but i guess we can't abort the civi.dao.preInsert without error or do an update on civi.dao.preInsert!!(Bad idea) and also want a way to alter the $params send to [create() function](https://github.com/civicrm/civicrm-core/blob/c9596e658d0e38727895b02f82de5ac7f149706e/CRM/Core/BAO/FinancialTrxn.php#L38)

**Comment by colemanw** [comment]
@pradpnayak this is a good change in the right direction, but I'd like to encourage you to go all the way since you're more than half-way there already, and finishing this one will get your head into the new way of writing BAO crud functionality.

The end-goal is to deprecate and eventually remove all BAO `create`, `add` and `del` functions. Most of those functions currently consist of:
1. Pre-save logic
2. Boilerplate code to copyValues & save
3. Post-save logic

As you correctly surmised, `#2` gets replaced with `self::writeRecord()`. The pre and post logic should be moved to hook callbacks (note, in order for a BAO to listen to hooks, you have to add ` implements \Civi\Test\HookInterface` to the class declaration). When finished, the entire `create()` function should be one line long (just calling `self::writeRecord()`) with a `@deprecated` annotation. I've already done this with a bunch of `del()` functions, breaking each single function into those 3 pieces. For example:

https://github.com/civicrm/civicrm-core/blob/11a1236cf36ca49443af1d644bdb42fd58ae180c/CRM/Case/BAO/CaseType.php#L434-L471

**Comment by eileenmcnaughton** [comment]
That test fail is REALLY hard to follow - hopefully jenkins (test this please) is just having a temporary spat....

I agree with the suggestions both by @pradpnayak and @colemanw - I don't think they are blocking on this PR & the change seems to make sense - so I would think jenkins is the main blocker on this PR as it stands

**Comment by pradpnayak** [comment]
Thanks @colemanw for the suggestion

@eileenmcnaughton, @colemanw 
Can i start with 

1. Move CRM/Core/BAO/FinancialTrxn.php to CRM/Financial/BAO/FinancialTrxn.
2. Replace class in all code from CRM_Core_BAO_FinancialTrxn to CRM_Financial_BAO_FinancialTrxn
3. Create empty file CRM/Core/BAO/FinancialTrxn.php which extends CRM_Financial_BAO_FinancialTrxn (not sure how to write deprecated warning!!!!)

**Comment by eileenmcnaughton** [comment]
@pradpnayak yep that sounds good - you probably realise you need to update the xml & regen

the deprecation notices are just `    CRM_Core_Error::deprecatedFunctionWarning('the right function');`

**Comment by jaapjansma** [comment]
@pradpnayak are you going to work on this? If not, no problem, but we will then advice to close this PR.

**Comment by eileenmcnaughton** [comment]
Let's close this & @pradpnayak can re-open when ready - that ALSO means that when a new PR is opened it will be at the top of the queue & get seen - old PRs that have been fixed up are often never noticed
